pulpcore/content/handler.py

```diff
@@ -215,42 +216,58 @@ def _permit(request, distribution):
 
+        If a fatal download failure occurs while downloading and there are additional
+        :class:`~pulpcore.plugin.models.RemoteArtifact` objects associated with the
+        :class:`~pulpcore.plugin.models.ContentArtifact` they will also be tried. If all
+        :class:`~pulpcore.plugin.models.RemoteArtifact` downloads raise exceptions, an HTTP 502
+        error is returned to the client.
+
         Args:
             request(:class:`~aiohttp.web.Request`): The request to prepare a response for.
             response (:class:`~aiohttp.web.StreamResponse`): The response to stream data to.
             content_artifact (:class:`~pulpcore.plugin.models.ContentArtifact`): The ContentArtifact
                 to fetch and then stream back to the client
-        """
-        remote_artifact = content_artifact.remoteartifact_set.get()
-        remote = remote_artifact.remote.cast()
-
-        async def handle_headers(headers):
-            for name, value in headers.items():
-                if name.lower() in HOP_BY_HOP_HEADERS:
-                    continue
-                response.headers[name] = value
-            await response.prepare(request)
-
-        async def handle_data(data):
-            await response.write(data)
-            if remote.policy != Remote.STREAMED:
-                await original_handle_data(data)
 
-        async def finalize():
+        Raises:
+            :class:`~aiohttp.web.HTTPNotFound` when no
+                :class:`~pulpcore.plugin.models.RemoteArtifact` objects associated with the
+                :class:`~pulpcore.plugin.models.ContentArtifact` returned the binary data needed for
+                the client.
+        """
+        for remote_artifact in content_artifact.remoteartifact_set.all():
+            remote = remote_artifact.remote.cast()
+
+            async def handle_headers(headers):
+                for name, value in headers.items():
+                    if name.lower() in HOP_BY_HOP_HEADERS:
+                        continue
+                    response.headers[name] = value
+                await response.prepare(request)
+
+            async def handle_data(data):
+                await response.write(data)
+                if remote.policy != Remote.STREAMED:
+                    await original_handle_data(data)
+
+            async def finalize():
+                if remote.policy != Remote.STREAMED:
+                    await original_finalize()
+
+            downloader = remote.get_downloader(remote_artifact=remote_artifact,
+                                               headers_ready_callback=handle_headers)
+            original_handle_data = downloader.handle_data
+            downloader.handle_data = handle_data
+            original_finalize = downloader.finalize
+            downloader.finalize = finalize
+            try:
+                download_result = await downloader.run()
+            except ClientResponseError:
```

Comment:
@bmbouter Sorry for the late (again) comment: Do you know what happens if the ClientResponseError occurs in an ongoing download (e.g. server closes connection unexpectedly)?

The code seems to look for the next remote artifact and put it into the same response (which may cause an exception for `response.prepare` or, even worse, just append the content to the broken content of the first artifact.

(I have not tested this, I may be wrong)

Reply:
I finally got around to filing this https://pulp.plan.io/issues/4434